src/OS8104A.ts

```diff
@@ -0,0 +1,697 @@
+import { getRegisterConfig } from "./RegisterConfig"
+import { Registers } from "./Registers.js"
+const TRANSFER_SPEED = 180000
+
+export interface MostMessage {
+    targetAddressHigh: number
+    targetAddressLow: number
+    fBlockID: number
+    instanceID: number
+    fktId: number
+    opType: number
+    data: number[]
+}
+
+export interface Config {
+    freq: number
+    nodeAddressLow: number
+    nodeAddressHigh: number
+    groupAddress: number
+}
+
+export interface Stream {
+    sourceAddrHigh: number
+    sourceAddrLow: number
+    fBlockID: number
+    instanceID: number
+    sinkNr: number
+}
+
+export interface AllocResult {
+    loc1: number
+    loc2: number
+    loc3: number
+    loc4: number
+    cl: number
+    answer1: string
+    freeChannels: number
+}
+
+export interface SourceResult {
+    nodePos: number
+    group: number
+    logicalHigh: number
+    logicalLow: number
+}
+
+const options: SpiOptions = {
+    chipSelectHigh: false,
+    bitsPerWord: 8,
+    lsbFirst: false
+}
+
+export class OS8104A extends EventEmitter {
+    readonly freq: number
+    readonly spi: SpiDevice
+    readonly interrupt: Gpio
+    readonly fault: Gpio
+    readonly status: Gpio
+    getRegisterConfig: typeof getRegisterConfig
+    mostStatus: Gpio
+    reset: Gpio
+    nodeAddressBuf: Buffer
+    groupAddressBuf: Buffer
+    awaitAlloc: boolean
+    allocResult: AllocResult
+    awaitGetSource: boolean
+    getSourceResult: SourceResult | null
+    allocTimeout: ReturnType<typeof setTimeout> | null
+    streamAllocTimeout: ReturnType<typeof setTimeout> | null
+    delayTimer: ReturnType<typeof setTimeout> | null
+    getSourceTimeout: ReturnType<typeof setTimeout> | null
+    allocCheck: ReturnType<typeof setInterval> | null
+    lockInterval: ReturnType<typeof setInterval> | null
+    multiPartMessage: MostMessage
+    multiPartSequence: number
+    transceiverLocked: boolean
+
+    constructor(nodeAddress: number, groupAddress: number, freq: number) {
+        super()
+        this.spi = spi.openSync(0, 0, options)
+        this.interrupt = new Gpio(5, "in", "falling")
+        // TODO this had an unnoticed type error for debounce, now TS has saved the day, it may mess things up
+        // now that it's actually working
+        this.fault = new Gpio(6, "in", "both", { debounceTimeout: 50 })
+        this.status = new Gpio(16, "in", "both", { debounceTimeout: 50 })
+        this.mostStatus = new Gpio(26, "in", "both", { debounceTimeout: 10 })
+        this.reset = new Gpio(17, "out")
+        this.freq = freq
+        // TODO not sure why these were buffers, need to review
+        this.nodeAddressBuf = Buffer.alloc(2)
+        this.nodeAddressBuf.writeUint16BE(nodeAddress)
+        this.groupAddressBuf = Buffer.alloc(1)
+        this.groupAddressBuf.writeUInt8(groupAddress)
+        this.awaitAlloc = false
+        // REVIEW - is this a better method than just allowing the type to be a type or null?
+        this.allocResult = {
```

Comment:
I personally prefer null or undefined - the reasoning is that this is not an actual allocation, the true state is that we did not allocate anything - therefore it in fact should be undefined at this point - you dont have to use null, but rather mark it as optional using ?

Reply:
done as above